src/main/java/duke/Parser.java

```diff
@@ -0,0 +1,187 @@
+
+    private void list(TaskList tasks) throws DukeException {
+        if (tasks.size() == 0) {
+            throw new DukeException("There are currently no tasks in your list.");
+        }
+        String matching = this.isFind ? "matching " : "";
+        StringBuilder tasksBuilder = new StringBuilder();
+        tasksBuilder.append("Here are the ").append(matching).append("tasks in your list:\n");
+        for (int i = 0; i < tasks.size(); ++i) {
+            String counter = String.valueOf(i + 1);
+            Task currentTask = tasks.get(i);
+            tasksBuilder.append(counter).append(".").append(currentTask);
+            if (i < tasks.size() - 1) {
+                // Append newline for all tasks before last task
+                tasksBuilder.append("\n");
+            }
+        }
+        Ui.printReply(tasksBuilder.toString());
+    }
+
+    private void done(int counter) throws DukeException {
+        if (counter <= 0 || counter > tasks.size()) {
+            throw new DukeException("Sorry, no such task of index " + counter + ".");
+        }
+        Task doneTask = tasks.get(counter - 1);
+        doneTask.markAsDone();
+        Ui.printReply("Nice! I've marked this task as done:\n  " + doneTask);
+    }
+
+    /**
+     * Getter for the rewrite status of the last parsed command.
+     *
+     * @return True if data in Storage needs to be rewritten.
+     */
+    public boolean toRewrite() {
+        return this.toRewriteData;
+    }
+
+    /**
+     * Getter for the exit status of the last parsed command.
+     *
+     * @return True if ready to exit Duke after "bye" command.
+     */
+    public boolean toExit() {
+        return this.isExit;
+    }
+
+    /**
+     * Parses the user input and runs the corresponding command.
+     *
+     * @param readIn A string containing the user input.
+     * @return A TaskList containing the updated Tasks after command is parsed.
+     * @throws DukeException A Duke-specific exception that occurs when user input is parsed.
+     */
+    public TaskList parse(String readIn) throws DukeException {
+        if (readIn.equals("bye")) {
+            Ui.printReply("Bye. Hope to see you again soon!");
+            this.isExit = true;
+        } else if (readIn.equals("list")) {
+            list(this.tasks);
+        } else {
+            String[] commandArguments = readIn.split(" ", 2);
+            String command = commandArguments[0];
+            String arguments = "";
+            if (commandArguments.length == 2) {
+                arguments = commandArguments[1];
+            }
+
+            switch (command) {
+            case "done": {
+                if (commandArguments.length < 2) {
+                    throw new DukeException("☹ OOPS!!! The index of '" + command + "' cannot be empty.");
+                }
+                this.toRewriteData = true;
+                int counter = Integer.parseInt(arguments);
+                done(counter);
+                break;
+            }
```

Comment:
As joel mentioned, might be good to stick to convention and not put {} around switch, case blocks